src/backend/InvenTree/part/tasks.py

```diff
@@ -77,6 +154,79 @@ def notify_low_stock_if_required(part_id: int):
             InvenTree.tasks.offload_task(notify_low_stock, p, group='notification')
 
 
+@tracer.start_as_current_span('check_stale_stock')
+@scheduled_task(ScheduledTask.DAILY)
+def check_stale_stock():
+    """Check all stock items for stale stock.
+
+    This function runs daily and checks if any stock items are approaching their expiry date
+    based on the STOCK_STALE_DAYS global setting.
+
+    For any stale stock items found, notifications are sent to users who have subscribed
+    to notifications for the respective parts. Each user receives one consolidated email
+    containing all their stale stock items.
+    """
+    # Sleep a random number of seconds to prevent worker conflict
+    time.sleep(random.randint(1, 5))
+
+    # Check if stock expiry functionality is enabled
+    if not get_global_setting('STOCK_ENABLE_EXPIRY', False, cache=False):
+        logger.info('Stock expiry functionality is not enabled - exiting')
+        return
+
+    # Check if STOCK_STALE_DAYS is configured
+    stale_days = int(get_global_setting('STOCK_STALE_DAYS', 0, cache=False))
+
+    if stale_days <= 0:
+        logger.info('Stock stale days is not configured or set to 0 - exiting')
+        return
+
+    today = InvenTree.helpers.current_date()
+    stale_threshold = today + timedelta(days=stale_days)
+
+    # Find stock items that are stale (expiry date within STOCK_STALE_DAYS)
+    stale_stock_items = stock_models.StockItem.objects.filter(
+        stock_models.StockItem.IN_STOCK_FILTER,  # Only in-stock items
+        expiry_date__isnull=False,  # Must have an expiry date
+        expiry_date__lt=stale_threshold,  # Expiry date is within stale threshold
+    ).select_related('part', 'location')  # Optimize queries
+
+    if not stale_stock_items.exists():
+        logger.info('No stale stock items found')
+        return
+
+    logger.info('Found %s stale stock items', stale_stock_items.count())
+
+    # Group stale stock items by user subscriptions
+    user_stale_items = {}  # Dictionary: user -> list of stale stock items
+
+    for stock_item in stale_stock_items:
+        # Get all subscribers for this part
+        subscribers = stock_item.part.get_subscribers()
+
+        for user in subscribers:
+            if user not in user_stale_items:
+                user_stale_items[user] = []
+            user_stale_items[user].append(stock_item)
+
+    # Send one consolidated notification per user
+    for user, items in user_stale_items.items():
+        try:
+            InvenTree.tasks.offload_task(
```

Comment:
I think that it would make sense to send a batch notification per stale item. The "trigger notification" method takes multiple users. This would be far more efficient

Reply:
@gitbock sorry, ignore that, I see that you are sending a single notification *per user* for the stale items